### ISSUE TITLE
server template with SRV record fills the backend with duplicate IP addresses during reloads

### ISSUE THREAD

**Issue by VigneshSP94** [haproxy/haproxy#1333]
### Detailed Description of the Problem

We are trying to using DNS SD in haproxy with server-templates to avoid reloads during updates to the backends, which works very well. 

But when we reload the haproxy service and let the state to be taken from a state-file, the backend gets filled with duplicate/stale ip addresses.

this is the SRV record queried by haproxy.
results in with one A record.
```
root@hapserver:~# dig +short @127.0.0.3 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local srv
10 1 83 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.local.
```
backend state is matching this result 
```
# be_id be_name srv_id srv_name srv_addr srv_op_state srv_admin_state srv_uweight srv_iweight srv_time_since_last_change srv_check_status srv_check_result srv_check_health srv_check_state srv_agent_state bk_f_forced_id srv_f_forced_id srv_fqdn srv_port srvrecord srv_use_ssl srv_check_port srv_check_addr srv_agent_addr srv_agent_port
20 v2::sg::_certtest::lb1::80::sg1 1 sg11 192.168.1.1 2 64 1 1 103 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.local 83 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 2 sg12 - 0 32 1 1 103 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 3 sg13 - 0 32 1 1 103 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 4 sg14 - 0 32 1 1 103 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 5 sg15 - 0 32 1 1 103 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 6 sg16 - 0 32 1 1 103 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 7 sg17 - 0 32 1 1 103 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 8 sg18 - 0 32 1 1 103 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 9 sg19 - 0 32 1 1 103 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 10 sg110 - 0 32 1 1 103 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83 
```

Now I'm reloading the service at 20 seconds interval for many times -
```
~# while true; do systemctl reload haproxy; sleep 20; done
# be_id be_name srv_id srv_name srv_addr srv_op_state srv_admin_state srv_uweight srv_iweight srv_time_since_last_change srv_check_status srv_check_result srv_check_health srv_check_state srv_agent_state bk_f_forced_id srv_f_forced_id srv_fqdn srv_port srvrecord srv_use_ssl srv_check_port srv_check_addr srv_agent_addr srv_agent_port
20 v2::sg::_certtest::lb1::80::sg1 1 sg11 192.168.1.1 2 0 1 1 450 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.local 83 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 2 sg12 192.168.1.1 2 0 1 1 450 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.local 83 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 3 sg13 192.168.1.1 2 0 1 1 450 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.local 83 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 4 sg14 192.168.1.1 2 0 1 1 450 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.local 83 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 5 sg15 192.168.1.1 2 0 1 1 450 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.local 83 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 6 sg16 192.168.1.1 2 0 1 1 450 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.local 83 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 7 sg17 192.168.1.1 2 0 1 1 450 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.local 83 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 8 sg18 192.168.1.1 2 0 1 1 450 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.local 83 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 9 sg19 - 0 32 1 1 450 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
20 v2::sg::_certtest::lb1::80::sg1 10 sg110 - 0 32 1 1 450 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local 0 83 - - 83
```

Backend gets filled with duplicate ip addresses and it stays that way forever until I remove the state file and restart the service.

### Expected Behavior

The backend should get updated to match what records received from DNS.

### Steps to Reproduce the Behavior

Create a server group with server-template and SRV record.
Let the state to be loaded from a state-file.
Reload several times.

### Do you have any idea what may have caused this?

No

### Do you have an idea how to solve the issue?

No

### What is your configuration?

```haproxy
global
  user haproxy
  group haproxy
  nbproc 1
  nbthread 16
  cpu-map auto:1/1-16 0-15
  log /dev/log local2
  log /dev/log local0 notice
  chroot /var/lib/haproxy
  pidfile /var/run/haproxy.pid
  daemon
  master-worker
  maxconn 200000
  hard-stop-after 1h
  stats socket /var/haproxy/stats mode 660 level admin expose-fd listeners
  tune.ssl.cachesize 3000000
  tune.ssl.lifetime 60000
  ssl-default-bind-ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256
  ssl-default-bind-options ssl-min-ver TLSv1.2 ssl-max-ver TLSv1.2
  server-state-file /tmp/states
  tune.bufsize 40960

defaults
  mode http
  log global
  retries 3
  timeout http-request 10s
  timeout queue 10s
  timeout connect 10s
  timeout client 1m
  timeout server 1m
  timeout tunnel 10m
  timeout client-fin 30s
  timeout server-fin 30s
  timeout check 10s
  option httplog
  option forwardfor except 127.0.0.0/8
  option redispatch
  load-server-state-from-file global

resolvers mydns
        nameserver dns1 stream+ipv4@127.0.0.3:53
        accepted_payload_size 8192 # allow larger DNS payloads
        timeout resolve 1s
        timeout retry   2s
        hold other    50000s
        hold refused  50000s
        hold timeout  50000s
        hold nx       5s
        hold valid    1s
        hold obsolete 1s


frontend one-listener
        bind x.x.x.x:80 mss 1440 alpn h2,http/1.1
        mode http
        option httplog
        default_backend one-backend

backend one-backend
        mode  http
        option tcp-check
        server-template sg1 10 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.local init-addr last,libc,none resolvers mydns check port 83  maxconn 3333 maxqueue 1 resolve-opts prevent-dup-ip

        errorfile 503 /etc/busy-html.http
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.2-553dee3 2021/07/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.2.html
Running on: Linux 4.15.0-42-generic #45-Ubuntu SMP Thu Nov 15 19:32:57 UTC 2018 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE=1 USE_LINUX_TPROXY=1 USE_LINUX_SPLICE=1 USE_LIBCRYPT=1 USE_OPENSSL=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL -LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=16).
Built with OpenSSL version : OpenSSL 1.1.1h  22 Sep 2020
Running on OpenSSL version : OpenSSL 1.1.1h  22 Sep 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.39 2016-06-14
Running on PCRE version : 8.39 2016-06-14
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 7.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace



Linux hapserver 4.15.0-42-generic #45-Ubuntu SMP Thu Nov 15 19:32:57 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by capflam** [comment]
I'm unable to reproduce the issue. Just a question. When HAProxy is reloaded, is the state file updated before the reload ?

**Comment by capflam** [comment]
In fact, I'm pretty sure I found the bug. I'll push a fix soon.

**Comment by VigneshSP94** [comment]
@capflam yes, we dump the state before reloads. can you try having multiple backends and try many reloads? I'm sure you will hit it. I just tried it, with only one backend, it hardly happens. Maybe try 20+ backends.

**Comment by VigneshSP94** [comment]
Also, I tried reading the code. I see there is a function written to check whether the IP address is already assigned to a server in the backend. But this function is not being called anywhere, it is just a function with no calls.

```
/*
 * Function to check if <ip> is already affected to a server in the backend
 * which owns <srv> and is up.
 * It returns a pointer to the first server found or NULL if <ip> is not yet
 * assigned.
 *
 * Must be called with server lock held
 */
struct server *snr_check_ip_callback(struct server *srv, void *ip, unsigned char *ip_family)
{
	struct server *tmpsrv;
	struct proxy *be;

	if (!srv)
		return NULL;

	be = srv->proxy;
	for (tmpsrv = be->srv; tmpsrv; tmpsrv = tmpsrv->next) {
		/* we found the current server is the same, ignore it */
		if (srv == tmpsrv)
			continue;

		/* We want to compare the IP in the record with the IP of the servers in the
		 * same backend, only if:
		 *   * DNS resolution is enabled on the server
		 *   * the hostname used for the resolution by our server is the same than the
		 *     one used for the server found in the backend
		 *   * the server found in the backend is not our current server
		 */
		HA_SPIN_LOCK(SERVER_LOCK, &tmpsrv->lock);
		if ((tmpsrv->hostname_dn == NULL) ||
		    (srv->hostname_dn_len != tmpsrv->hostname_dn_len) ||
		    (strcasecmp(srv->hostname_dn, tmpsrv->hostname_dn) != 0) ||
		    (srv->puid == tmpsrv->puid)) {
			HA_SPIN_UNLOCK(SERVER_LOCK, &tmpsrv->lock);
			continue;
		}

		/* If the server has been taken down, don't consider it */
		if (tmpsrv->next_admin & SRV_ADMF_RMAINT) {
			HA_SPIN_UNLOCK(SERVER_LOCK, &tmpsrv->lock);
			continue;
		}

		/* At this point, we have 2 different servers using the same DNS hostname
		 * for their respective resolution.
		 */
		if (*ip_family == tmpsrv->addr.ss_family &&
		    ((tmpsrv->addr.ss_family == AF_INET &&
		      memcmp(ip, &((struct sockaddr_in *)&tmpsrv->addr)->sin_addr, 4) == 0) ||
		     (tmpsrv->addr.ss_family == AF_INET6 &&
		      memcmp(ip, &((struct sockaddr_in6 *)&tmpsrv->addr)->sin6_addr, 16) == 0))) {
			HA_SPIN_UNLOCK(SERVER_LOCK, &tmpsrv->lock);
			return tmpsrv;
		}
		HA_SPIN_UNLOCK(SERVER_LOCK, &tmpsrv->lock);
	}


	return NULL;
}
```

![image](https://user-images.githubusercontent.com/24570763/126610587-7e04e073-c763-4742-ac2e-0b7d7b222b63.png)

**Comment by capflam** [comment]
I pushed a fix in upstream. Could you confirm it fixes your issue ? The commit can safely be applied on top of the 2.4.2 or 2.4-HEAD. Otherwise, you can wait a bit. The fix will be backported soon.

**Comment by VigneshSP94** [comment]
@capflam I cloned the upstream master and tried the same, I don't see the issue happening again. No duplicate values.

But now I see a different issue -

server-state matches the DNS response

```
root@hapserver:~# dig +short @127.0.0.3 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local srv
10 1 83 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local.

1
# be_id be_name srv_id srv_name srv_addr srv_op_state srv_admin_state srv_uweight srv_iweight srv_time_since_last_change srv_check_status srv_check_result srv_check_health srv_check_state srv_agent_state bk_f_forced_id srv_f_forced_id srv_fqdn srv_port srvrecord srv_use_ssl srv_check_port srv_check_addr srv_agent_addr srv_agent_port
20 one-backend 1 sg11 100.93.80.5 2 0 1 1 12 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 83 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 83
20 one-backend 2 sg12 - 0 32 1 1 12 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 83
20 one-backend 3 sg13 - 0 32 1 1 12 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 83
20 one-backend 4 sg14 - 0 32 1 1 12 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 83
20 one-backend 5 sg15 - 0 32 1 1 12 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 83
20 one-backend 6 sg16 - 0 32 1 1 12 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 83
20 one-backend 7 sg17 - 0 32 1 1 12 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 83
20 one-backend 8 sg18 - 0 32 1 1 12 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 83
20 one-backend 9 sg19 - 0 32 1 1 12 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 83
20 one-backend 10 sg110 - 0 32 1 1 12 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 83
```

The `srv_check_port` in the state is taken from the `check port` of `server-template`.

```
backend one-backend
        mode  http
        option tcp-check
        server-template sg1 10 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local init-addr last,libc,none resolvers mydns check port 83  maxconn 10000 maxqueue 1 resolve-opts prevent-dup-ip
```

Now I change the port in DNS SRV record and `check port` of the config as well. DNS response gives 84 as port and `check port` reads same value.

```
root@hapserver:~# dig +short @127.0.0.3 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local srv
10 1 84 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local.

backend one-backend
        mode  http
        option tcp-check
        server-template sg1 10 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local init-addr last,libc,none resolvers mydns check port 84  maxconn 10000 maxqueue 1 resolve-opts prevent-dup-ip
```

Now checking the state of this backend, it picks up the listen port from DNS perfectly fine, but the `check port` value is stuck in the previous value. It doesn't get the update from file after a reload.

```
1
# be_id be_name srv_id srv_name srv_addr srv_op_state srv_admin_state srv_uweight srv_iweight srv_time_since_last_change srv_check_status srv_check_result srv_check_health srv_check_state srv_agent_state bk_f_forced_id srv_f_forced_id srv_fqdn srv_port srvrecord srv_use_ssl srv_check_port srv_check_addr srv_agent_addr srv_agent_port
20 one-backend 1 sg11 - 0 32 1 1 6 6 3 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 84
20 one-backend 2 sg12 100.93.80.5 2 64 1 1 184 6 3 4 6 0 0 0 my-server01.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 84 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 84
20 one-backend 3 sg13 - 0 32 1 1 184 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 84
20 one-backend 4 sg14 - 0 32 1 1 184 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 84
20 one-backend 5 sg15 - 0 32 1 1 184 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 84
20 one-backend 6 sg16 - 0 32 1 1 184 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 84
20 one-backend 7 sg17 - 0 32 1 1 184 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 84
20 one-backend 8 sg18 - 0 32 1 1 184 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 84
20 one-backend 9 sg19 - 0 32 1 1 184 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 84
20 one-backend 10 sg110 - 0 32 1 1 184 1 0 0 14 0 0 0 - 0 _servers._tcp.zone.611c3e38786a57b4337a588b8cad9020-80.lbaas.local 0 83 - - 84
```

Is my expectation to see it changed to what I have in file correct?

**Comment by capflam** [comment]
`srv_check_port` is set using the configuration value but it may be overridden by the value found in the server state file, if defined. The DNS resolution does not change the check port. To change it dynamically, you should set it via the CLI using `set server` command. However, if server and health-check port are always the same, there is no reason to set `port` option on the server line. The server port is used by default for health-check.

**Comment by VigneshSP94** [comment]
@capflam thank you, understood.

The commit you provided in the upstream does fix the issue, will this be available in any release soon?

**Comment by capflam** [comment]
I will first backport the fix as far as 2.2. No release is planned for now. We must first check the pending fixes. But the 2.4 will probably be released soon (next week or so).

**Comment by VigneshSP94** [comment]
@capflam thank you, we will wait for next 2.4x release.

**Comment by FLavalliere** [comment]
@capflam  would this be the reason why i could of had the following issue:

Haproxy version: 2.4.2-553dee3 

server-template xxxx.ca 4 .....

But in the 'show servers state' or the state file have 5 entries?

```
1042 xxxx.ca 1 xxxx.ca1 10.10.12.83 0 0 1 1 1370 8 2 0 6 0 0 0 database-arm-v11-03e55.node.aws-us-east-1.consul 32813 _tcp_.entrypoint.4c3aeb70e31c11e994c6cd213920fbdc.service.consul 0 0 - - 0
1042 xxxx.ca 2 xxxx.ca2 10.10.12.83 0 0 1 1 1370 8 2 0 6 0 0 0 database-arm-v11-03e55.node.aws-us-east-1.consul 32813 _tcp_.entrypoint.4c3aeb70e31c11e994c6cd213920fbdc.service.consul 0 0 - - 0
1042 xxxx.ca 3 xxxx.ca3 10.10.12.83 0 0 1 1 1370 8 2 0 6 0 0 0 database-arm-v11-03e55.node.aws-us-east-1.consul 32813 _tcp_.entrypoint.4c3aeb70e31c11e994c6cd213920fbdc.service.consul 0 0 - - 0
1042 xxxx.ca 4 xxxx.ca4 10.10.12.83 0 0 1 1 1370 8 2 0 6 0 0 0 database-arm-v11-03e55.node.aws-us-east-1.consul 32813 _tcp_.entrypoint.4c3aeb70e31c11e994c6cd213920fbdc.service.consul 0 0 - - 0
1042 xxxx.ca 5 https_nobackend 127.0.0.1 2 0 1 1 33430 1 0 3 0 0 0 0 - 439 - 0 0 - - 0
```

Since the server-template is supposed to only read the first 4 records, i and somehow the one that resolved properly is at the 5th entry then this returned an 503 error...

**Comment by FLavalliere** [comment]
Actually, its 3 duplicate entries i see... so looks like its probably related. I rebuilded using the upstream fixes (from 2.4 master branch) will see if it  re-occurs.

**Comment by capflam** [comment]
@FLavalliere, yes it seems to be the same issue. But you must load servers state from a file using `server-state-file-name` or `server-state-file` to hit the bug.

**Comment by FLavalliere** [comment]
I guess we do:

```
  server-state-file state
  server-state-base /var/state/haproxy/

defaults
  load-server-state-from-file global
```

Is this still required to load servers from state file with server-template for hot-reloads? 

The reason we used the server-state was for hot-reloading (with server-template) was to ensure the haproxy doesnt have 'down' servers while reloading.

**Comment by VigneshSP94** [comment]
@capflam is this fix available in any 2.4 releases?

**Comment by capflam** [comment]
Not yet, but a new 2.4 release is planned this week.
